### PR TITLE
[M1839] Update summary models in batches

### PR DESCRIPTION
**Pull request by gridcell**
In order to reduce memory usage during update summary models are updated in batches.  SQLModel  is filtered by sample event ids (10 sample events at a time).

### CODE REVIEW

**Comment by gridcell** [comment]
[Reduce RAM usage when updating summary caches](https://trello.com/c/FOGlCh1l/1839-reduce-ram-usage-when-updating-summary-caches)

File: src/api/models/sql_models/base.py [hunk 1]
```diff
@@ -6,7 +6,9 @@
 from api.models import Project
 
 
-project_where = """    WHERE\n        project.id = '%(project_id)s' :: uuid\n"""
+project_where = """project.id = '%(project_id)s' :: uuid"""
+sample_event_where = """se.id::uuid = ANY(ARRAY[%(sample_event_id)s]::uuid[])"""
+# sample_event_where = """se.id = '%(sample_event_id)s' :: uuid"""
```
Comment (saanobhaai) [inline src/api/models/sql_models/base.py:11]: cleanup comment

File: src/api/models/sql_models/base.py [hunk 2]
```diff
@@ -6,7 +6,9 @@
 from api.models import Project
 
 
-project_where = """    WHERE\n        project.id = '%(project_id)s' :: uuid\n"""
+project_where = """project.id = '%(project_id)s' :: uuid"""
```
Comment (saanobhaai) [inline src/api/models/sql_models/base.py:9]: ok but the spacing was actually nice when printing out for testing :-)

File: src/sqltables/query.py [hunk 3]
```diff
@@ -129,6 +129,8 @@ def __init__(self, model=None, query=None, using=None, hints=None):
         self.query = query or SQLTableQuery(self.model)
 
     def sql_table(self, **sql_table_params: Dict[str, Any]) -> "SQLTableQuerySet":
+        self.query.model_sql = self.query.model_sql.replace("<<__sql_table_args__>>", " AND ".join(["1 = 1", *[self.model.sql_args[field].sql for field  in sql_table_params]]))
```
Comment (saanobhaai) [inline src/sqltables/query.py:132]: Nice

File: src/api/utils/summary_cache.py [hunk 4]
```diff
@@ -54,54 +56,81 @@
...
+    for record in records:
+        record.created_on = created_on
+
+
+def _update_records(project_id, records, target_model_cls, created_on):
     target_model_cls.objects.filter(project_id=project_id).delete()
     idx = 0
     while True:
         batch = records[idx : idx + BATCH_SIZE]
+        _set_created_on(created_on, batch)
```
Comment (saanobhaai) [inline src/api/utils/summary_cache.py:69]: Does this need to be a separate function?

File: src/api/utils/summary_cache.py [hunk 5]
```diff
@@ -54,54 +56,81 @@
...
         target_model_cls.objects.bulk_create(batch, batch_size=BATCH_SIZE)
         idx += BATCH_SIZE
 
 
-def _fetch_records(sql_model_cls, project_id):
-    return list(sql_model_cls.objects.all().sql_table(project_id=project_id))
+def _fetch_records(sql_model_cls, project_id, sample_event_ids=None):
+    return list(
+        sql_model_cls.objects.all().sql_table(
+            project_id=project_id, sample_event_id=sample_event_ids
```
Comment (saanobhaai) [inline src/api/utils/summary_cache.py:79]: Should `sql_table` arg be plural too, i.e. `sample_event_ids`?

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Brilliant.

**Pull request merged**